From 9f4d7a52a0f44541c299b08e0f805ff533f69918 Mon Sep 17 00:00:00 2001
From: Pavan Kumar Ramayanam <pavan.ramayanam@amd.com>
Date: Wed, 6 Dec 2017 17:16:50 +0530
Subject: [PATCH] drm/amdkfd: Fix a bug that process cleanup is not done
 properly

When destroying a queue fails in nocpsch, it may leave other queues
not destroyed and process not unregistered. This commit fixes that.

Change-Id: I9dcf7cb71242aa02e5b72c0381fec19a85f3dfe4

Signed-off-by: Pavan Kumar Ramayanam <pavan.ramayanam@amd.com>
---
 drivers/gpu/drm/amd/amdkfd/kfd_device_queue_manager.c | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/drm/amd/amdkfd/kfd_device_queue_manager.c b/drivers/gpu/drm/amd/amdkfd/kfd_device_queue_manager.c
index e303bae..7b25969a 100644
--- a/drivers/gpu/drm/amd/amdkfd/kfd_device_queue_manager.c
+++ b/drivers/gpu/drm/amd/amdkfd/kfd_device_queue_manager.c
@@ -1431,9 +1431,11 @@ static int process_termination_nocpsch(struct device_queue_manager *dqm,
 
 	/* Clear all user mode queues */
 	list_for_each_entry_safe(q, next, &qpd->queues_list, list) {
-		retval = destroy_queue_nocpsch_locked(dqm, qpd, q);
-		if (retval)
-			goto out;
+		int ret;
+
+		ret = destroy_queue_nocpsch_locked(dqm, qpd, q);
+		if (ret)
+			retval = ret;
 	}
 
 	/* Unregister process */
@@ -1446,7 +1448,6 @@ static int process_termination_nocpsch(struct device_queue_manager *dqm,
 		}
 	}
 
-out:
 	mutex_unlock(&dqm->lock);
 	return retval;
 }
-- 
2.7.4

